pkg/apis/resource/validation/validation.go

```diff
@@ -138,31 +169,81 @@ func gatherAllocatedDevices(allocationResult *resource.DeviceAllocationResult) s
+			},
+			func(subRequest resource.DeviceSubRequest) (string, string) {
+				return subRequest.Name, "name"
+			},
+			fldPath.Child("firstAvailableOf"))...)
+		return allErrs
 	}
-	allErrs = append(allErrs, validateSlice(request.Selectors, resource.DeviceSelectorsMaxSize,
-		func(selector resource.DeviceSelector, fldPath *field.Path) field.ErrorList {
-			return validateSelector(selector, fldPath, stored)
-		},
-		fldPath.Child("selectors"))...)
-	switch request.AllocationMode {
+	allErrs = append(allErrs, validateDeviceClass(request.DeviceClassName, fldPath)...)
+	allErrs = append(allErrs, validateSelectorSlice(request.Selectors, fldPath, stored)...)
+	allErrs = append(allErrs, validateDeviceAllocationMode(request.AllocationMode, request.Count, fldPath)...)
+	return allErrs
+}
+
+func validateDeviceSubRequest(subRequest resource.DeviceSubRequest, fldPath *field.Path, stored bool) field.ErrorList {
+	allErrs := validateRequestName(subRequest.Name, fldPath.Child("name"))
+	allErrs = append(allErrs, validateDeviceClass(subRequest.DeviceClassName, fldPath)...)
+	allErrs = append(allErrs, validateSelectorSlice(subRequest.Selectors, fldPath, stored)...)
+	allErrs = append(allErrs, validateDeviceAllocationMode(subRequest.AllocationMode, subRequest.Count, fldPath)...)
+	return allErrs
+}
+
+func validateDeviceAllocationMode(deviceAllocationMode resource.DeviceAllocationMode, count int64, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	switch deviceAllocationMode {
 	case resource.DeviceAllocationModeAll:
-		if request.Count != 0 {
-			allErrs = append(allErrs, field.Invalid(fldPath.Child("count"), request.Count, fmt.Sprintf("must not be specified when allocationMode is '%s'", request.AllocationMode)))
+		if count != 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("count"), count, fmt.Sprintf("must not be specified when allocationMode is '%s'", deviceAllocationMode)))
 		}
 	case resource.DeviceAllocationModeExactCount:
-		if request.Count <= 0 {
-			allErrs = append(allErrs, field.Invalid(fldPath.Child("count"), request.Count, "must be greater than zero"))
+		if count <= 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("count"), count, "must be greater than zero"))
 		}
 	default:
-		allErrs = append(allErrs, field.NotSupported(fldPath.Child("allocationMode"), request.AllocationMode, []resource.DeviceAllocationMode{resource.DeviceAllocationModeAll, resource.DeviceAllocationModeExactCount}))
+		allErrs = append(allErrs, field.NotSupported(fldPath.Child("allocationMode"), deviceAllocationMode, []resource.DeviceAllocationMode{resource.DeviceAllocationModeAll, resource.DeviceAllocationModeExactCount}))
+	}
+	return allErrs
+}
+
+func validateDeviceClass(deviceClass string, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if deviceClass == "" {
+		allErrs = append(allErrs, field.Required(fldPath.Child("deviceClassName"), ""))
```

Comment:
I prefer moving this `fldPath.Child("deviceClassName")` into the caller. Then `validateDeviceClass` can be used regardless of what the field is called.

I understand that this leads to more code duplication, but it's consistent with how it is usually done.

Same for all other "common" fields.